### ISSUE TITLE
Umbrella issue for reconciliation semantics

### ISSUE THREAD

**Issue by ash2k** [kubernetes-incubator/service-catalog#1001]
Fist of all I'm sorry if this is totally inappropriate or invalid :) It started as a comment  https://github.com/kubernetes-incubator/service-catalog/pull/979#discussion_r125159846 but became too big and I decided to create an issue instead. Feel free to close it/etc.

IMHO we should have a document that describes ideal/desired reconciliation behaviour in detail, ignoring what is possible with current OSB spec. Then, once we have it, we can take a look at the spec and see how we can implement things. It's not hard to imagine we will have behaviours that are hard/impractical/impossible to implement given the current OSB spec. That is how we will get drivers/use cases to improve the spec.
Building the other way around (here is the spec, lets build an impl) is certainly possible, but it seems to me top down approach will result in better native kubernetes-like experience.
Some things from the list below have tickets already. This is my attempt to answer/ask relevant questions and some learnings from implementing Smith.

### Re. reprocessing:
- When `Broker` is updated/added it should be queried for new list of `ServiceClasses`. Existing related `ServiceClasses` should be updated, new ones added, removed ones deleted.
- When `ServiceClass` is updated/delete/added all related `Instances` should be reprocessed.
- When `Instance` is updated/delete/added all related `Bindings` should be reprocessed.
- When `Binding` is updated/delete/added all related `Secrets` should be reprocessed.
- Reverse reprocessing: `ServiceClass`->`Broker` and `Secret`->`Binding` -  E.g. if something modifies the contents of a `ServiceClass` object we should fix it by reprocessing related `Broker`. If `Secret` is deleted/etc, the related `Binding` should be reprocessed so that `Secret` is recreated.

### Re. owner references:
Controller with "block owner deletion" flag set:
- `ServiceClass`->`Broker`
- `Secret`->`Binding`

Non-controller with "block owner deletion" flag set:
- `Instance`->`ServiceClass`
- `Binding`->`Instance`

### Re. deletion:
Deletion is tricky. Cascading deletion or orphaning?
- What happens when `Broker` is deleted? `ServiceClasses` should be deleted? What if there are `Instances`?
- What happens when `ServiceClass` is deleted? `Instances` should be deleted? What if there are `Bindings`?
- What happens when `Instance` is deleted? `Bindings` should be deleted.
- What happens when `Binding` is deleted? `Secret` should be deleted.

How deletion works depends on GC, owner references and deletion options (foreground, background vs orphaning).

### Re. adoption:
What if there is an orphaned `ServiceClass`/`Instance`/`Binding`/`Secret` and an owner is created. E.g. orphan `Secret` and a `Binding` is created. Should it adopt the `Secret`? Think about each combination.

### Re. blocked creation:
What if a `Broker` wants to create a `ServiceClass` or `Binding` wants to create a `Secret` and that object already exists? A terminal error, prehaps. However what if that conflicting object is deleted? The `Broker`/`Binding` should be reprocessed because it is unblocked.

Basically, this is the list of kinds of issues I faced with Smith so far :) All are solved via event listeners and additional indexes on informers.

### Mandatory reading:
https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md

Some pre-existing issues (missed something?): #780, #750, #705, #704, #601, #481, #801, #978, #999, #976, #921

**Comment by nilebox** [comment]
I agree that we need to follow the "Kubernetes native" approach for reconciliation loop whenever it is possible. However, I'm not sure if top-down approach will always work, as OSB was not designed for idempotent reprocessing.

> should be reprocessed

@ash2k we need a definition of "reprocess" there. Given that OSB spec currently doesn't have mandatory `GET` endpoints for `Instances` and `Bindings`, it's a non-trivial question.

The only reliable way of "fully reprocessing" in OSB is to delete the old object and create a new one. But even that is often not feasible - we can't simply wipe the data and create a new empty `Instance` for every update of `ServiceClass`.

**Comment by ash2k** [comment]
> The only reliable way of "fully reprocessing" in OSB is to delete the old object and create a new one. But even that is often not feasible - we can't simply wipe the data and create a new empty Instance for every update of ServiceClass.

@nilebox Sure. Well, if user has requested an update that cannot be fulfilled in principle then that is a terminal error. Easy :) I understand what you are talking about.

**Comment by nilebox** [comment]
> user has requested an update

@ash2k the problem here is that in Kubernetes' `Instance` object you only have the "desired state". To know exactly _what_ has changed (to determine whether an update _can_ be fulfilled or not), you need the snapshot of the previous version, or, a way to retrieve the "actual state" from the broker.

Hence the `GET` endpoint I mentioned, which could have solved this problem by providing an actual state, but since it's not required by the spec, we can't rely on that.

The best we can do (and I have already seen the code about this in Service Catalog) is to store the checksum of the entire `Instance` spec. This way we can at least know if the spec has changed since we've sent it to the broker.

>  then that is a terminal error

I am afraid that with such limitations any update could become a terminal error :(

**Comment by nilebox** [comment]
Well, at least we have `UPDATE` [for Instances](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#updating-a-service-instance) (with optional `plan_updateable` flag), but still no way to update the Binding.

My initial point was that the devil is in the detail, and we need to be explicit about what "reprocess" means for _every type_ of Kubernetes resource, and which changes are supported.
